Port zone verification from CreDNS to NSD4

k0ekk0ek opened the pull request:
This small changeset ports zone verification from CreDNS, which had it's last release in 2012(?), to NSD4.

The most significant changes are the ones to the transfer daemon (xfrd). xfrd will now keep zone transfers around until reload succeeds so that updates do not have to fetched again if verification for a single zone fails. Apart from that I've updated `nsd-control` to check an environment variable called `NSD_PATH` so the tests that came with CreDNS (@wtoorop's changes included) can simply be run from the build root.

The biggest change to the zone verification feature itself is that it now is implemented on top of libevent as opposed to netio. I'm sure there's more goodness that I forgot to mention, but only ~8600 lines changed so you should pick those up quickly :sweat_smile:.

Commit:
Use wait instead of sleep in xfr_udp to handle timing difference

Commit:
Add xfr_retry test to verify xfrd changes somewhat

Commit:
Port zone verification from CreDNS to NSD

Add support for verification of zones on updates. Allows the user to execute a
script or program when a zone is updated to verify if the contents is correct.
The verifier can keep the zone from being published if for example the zone was
incorrectly signed.

Commit:
Add verify test to test environment, zone feed and axfr

Commit:
Port zone verification tests from CreDNS to NSD

Commit:
Make nsd-control read envvar

Commit:
Add verify_timeout test

Commit:
Add verify_no test

wcawijngaards commented on configlexer.lex:
```diff
@@ -299,6 +299,16 @@ tls-service-key{COLON}	{ LEXOUT(("v(%s) ", yytext)); return VAR_TLS_SERVICE_KEY;
 tls-service-pem{COLON}	{ LEXOUT(("v(%s) ", yytext)); return VAR_TLS_SERVICE_PEM;}
 tls-port{COLON}		{ LEXOUT(("v(%s) ", yytext)); return VAR_TLS_PORT;}
+tls-service-pem{COLON}	{ LEXOUT(("v(%s) ", yytext)); return VAR_TLS_SERVICE_PEM;}
+tls-port{COLON}		{ LEXOUT(("v(%s) ", yytext)); return VAR_TLS_PORT;}
```
These two lines are duplicated.

k0ekk0ek replied:
When I switch workspace sporadically a line is duplicated in my IDE, thanks for catching this, I've removed them.

wcawijngaards commented on doc/README:
```diff
@@ -767,6 +768,102 @@ entirely on the hardware. cpu-affinity options are supported on Linux and
+	   with the "verifier-feed-zone:" option.
+
+	   Examples for verifiers that read from the standard input are:
+	   "lds-verify-zone -V2" (-V2 to suppress copying to stdout) or
```
s/lds/ldns/

k0ekk0ek replied:
Done

wcawijngaards commented on nsd-control.c:
```diff
@@ -521,8 +522,11 @@ int main(int argc, char* argv[])
 		usage();
 	if(argc >= 1 && strcmp(argv[0], "start")==0) {
-		if(execl(NSD_START_PATH, "nsd", "-c", cfgfile, 
-			(char*)NULL) < 0) {
+		const char *path;
+		if((path = getenv("NSD_PATH")) == NULL) {
+			path = NSD_START_PATH;
+		}
```
The new NSD_PATH env variable, I do not recall documentation for it (or maybe it is somewhere else), and that would be good to have.

k0ekk0ek replied:
It's needed solely to run the CreDNS test-suite that I copied and didn't document it on purpose to keep it "hidden". If I document it, it's a "supported" feature. Do we want that?

wcawijngaards replied:
Some hidden debug feature in our production code is unwanted.  If we can turn it into a supported feature that is useful for users, that would be good too.  Otherwise change the test setup (maybe just set $PATH or start nsd directly?)

wcawijngaards replied:
I mean, to answer more directly, I don't mind this as a supported feature.

k0ekk0ek replied:
It starts `nsd` directly. `NSD_START_PATH` is defined in `config.h`, that's the problem when trying to run in the build directory (I don't want to install while developing). Guess I'll be documenting the environment variable :slightly_smiling_face:. Should I add it to the manpage for `nsd-control`?

wcawijngaards replied:
I mean, it is possible to start nsd by starting ../../nsd instead of ../../nsd-control start.  Then the NSD_START_PATH does not matter, right?  But adding it to the manpage is what I originally asked you for.

k0ekk0ek replied:
Done

wcawijngaards commented on nsd.conf.sample.in:
```diff
@@ -248,6 +248,38 @@ server:
+	# port: 5347
+
+	# Interfaces to bind for zone verification (default are the wildcard
+	# interfaces 0.0.0.0 and ::0). To bind to to multiple IP addresses,
```
Default of 127.0.0.1 and ::1 is more secure.

k0ekk0ek replied:
Actually changed to code to pass "localhost" so it'll work as expected and open both IPv4 and IPv6 sockets.

wcawijngaards commented on nsd.h:
```diff
@@ -26,6 +26,7 @@ struct daemon_remote;
 struct dt_collector;
 #endif
+struct nsd_verifier_proxy;
 
```
It looks like the `struct nsd_verifier_proxy` is not used. If so, you can delete this line.

k0ekk0ek replied:
A remnant from a previous work. Thanks for catching this.

wcawijngaards commented on options.c:
```diff
@@ -1169,6 +1195,49 @@ unmarshal_acl_list(region_type* r, struct buffer* b)
+	for(i = 0; strv[i] != NULL; i++) {
+		marshal_str(b, strv[i]);
+	}
+	marshal_u8(b, 0);
```
This marshal_u8 does not match with the unmarshal_strv routine.  That routine reads u32 and then a list of strings.  So that is what this routine should write, but it adds an unused zero byte.

wcawijngaards replied:
Sorry now I see that this adds the NULL at end of the array.

wtoorop commented on nsd.c:
```diff
@@ -114,12 +114,68 @@ version(void)
+			setenv("VERIFY_IPV6_ADDRESS", host, 1);
+			setenv("VERIFY_IPV6_PORT", serv, 1);
+			setenv("VERIFY_IP_ADDRESS", host, 1);
+			setenv("VERIFY_IP_PORT", serv, 1);
```
In doc/README the environment variable is `VERIFY_PORT` instead of `VERIFY_IP_PORT`.  `VERIFY_PORT` is also used in `tpkg/credns-setup` and in `tpkg/credns.tdir` (although that also sets the variable in the main environment, which is probably the reason you didn't notice it).
I have no strong opinion whether this should be `VERIFY_PORT` or whether the other environment variables should be `VERIFY_IP_PORT`, but the number of occurrences of `VERIFY_IP_PORT` is much lower, so probably easier to change.

k0ekk0ek replied:
Good catch! Changed it back to `VERIFY_PORT`. Probably friendlier to users who want to migrate from CreDNS to NSD4 too :+1:

wtoorop commented:
@k0ekk0ek I am doing some tests with `tpkg/credns-setup` and noticing some issues.
When testing with just a single zone:
The verifier taking longer to verify than the SOA Retry timer, results in a constant stream of AXFR queries from the proxy to the master (which are luckily aborted by the querier/proxy):
```
[2020-04-06 11:50:10.903] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
[2020-04-06 11:50:29.921] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
[2020-04-06 11:50:47.938] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
[2020-04-06 11:51:06.957] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
[2020-04-06 11:51:25.973] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
[2020-04-06 11:51:44.991] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
[2020-04-06 11:52:03.009] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
```
This constant stream of AXFR queries eventually stops when the SOA Expire timer runs out.

Maybe, when the SOA Retry timer goes, we have to check whether the zone is still being verified and then postpone for another SOA Retry seconds?

k0ekk0ek commented:
@wtoorop, that's probably because xfrd is continuing? The transfer daemon shouldn't retry until there's an answer from the reload process really, so this is definitely a bug then.

wtoorop committed:
Change occurences of VERIFY_IP_PORT into VERIFY_PORT

wcawijngaards commented on server.c:
```diff
@@ -2786,6 +2843,114 @@ add_tcp_handler(
+	nsd->event_base = NULL;
+	nsd->server_region = NULL;
+	assert(nsd->next_zone_to_verify == NULL);
+	assert(nsd->verifier_count == 0);
```
The two assertions should be above the 'fail:' label.  They are not met when fail is used.  But they are met when a successful exit is performed.

k0ekk0ek replied:
Indeed. Good catch.

wtoorop committed:
Fix cutest for popen3 using file descriptors

wtoorop commented:
@k0ekk0ek I'm doing PRs on your branch for you to process (and merge if you like them).
All PR's have master already merged. Safes you some work. There are currently 2 PR's.
1. k0ekk0ek/nsd#2 : Rename VERIFY_IP_PORT into VERIFY_PORT
2. k0ekk0ek/nsd#3 : Fix cutest program to use fd's with popen3
That last PR should make the cirrus test succeed...

wcawijngaards commented on tpkg/common.sh:
```diff
@@ -250,6 +250,33 @@ wait_server_up_or_fail () {
+# $5: # times to try (# seconds dig is ran)
+wait_for_soa_serial () {
+	TS_START=`date +%s`
+	for i in `eval echo {1..$5}`
```
The `eval` call is not needed here, it is better not to use it, ``` for i in `echo {1..$5}` ```

wcawijngaards replied:
For portability to non-bash, this should be ``` for i in `seq 1 $5` ```

k0ekk0ek replied:
Changed it

wcawijngaards commented on tpkg/concurrent_credns.tdir/concurrent_credns.post:
```diff
@@ -0,0 +1,11 @@
+#
+# do your teardown here
+
+make realclean
```
Missing is `make stop` to stop the started NSD instances

k0ekk0ek replied:
~I guess that goes for `tpkg/credns.tdir/credns.post` to then~. I'll add it.

k0ekk0ek replied:
Not the case for the `credns` test. There it's located in `credns.test` near the bottom, but I think I'll move that one to the `post` script too. Better to stop the servers whether the test fails or succeeds.

k0ekk0ek replied:
Done

wtoorop committed:
Configuration has changed

wtoorop commented:
Another observation. This affects a non-verifying slave too.

When a slave is restarted, and it does not have the slave-zone saved to disk and no database configured, most of the times (but not always) it thinks it has an up-to-date zone loaded. `nsd-control zonestatus` will show an up to date zone.  Queries for the zone result in `SERVFAIL,`. Only `nsd-control force_transfer` fixes the situation then.

I tested and this issue was not present before this PR. Before the PR secondaries would transfer the zone if they did not have it on disk.

wtoorop committed:
Create control shortcuts

k0ekk0ek commented:
Offline comment by @wtoorop:

Second issue (secondary not xfr'ing the zone) is new in zone verification patches. So two issues:
1. AXFR flood with retry timer
2. Secondary does not xfr zone after restart if they've been successfully xfr'ed before.

wcawijngaards commented on tpkg/concurrent_credns.tdir/concurrent_credns.test:
```diff
@@ -0,0 +1,68 @@
+make start-proxy
+wait_nsd_up proxy/nsd.log
+make start-master
+wait_nsd_up proxy/nsd.log
```
The last one is supposed to be wait_nsd_up for master/nsd.log, instead of proxy/nsd.log, I think?

k0ekk0ek replied:
Would make sense...

k0ekk0ek replied:
Done

wcawijngaards commented on tpkg/credns-setup/Makefile.in:
```diff
@@ -0,0 +1,236 @@
+     || echo "-- Could not (re)start $(SLAVE)"
+
+start: start-$(MASTER) start-$(PROXY) start-$(SLAVE)
+restart: restart-$(MASTER) restart-$(PROXY) restart-$(SLAVE)
```
The restart-* commands do not exist. The start-* ones are defined, also with 'restart' echo, but the line 'restart: ...' is not operative.  I guess you could remove it if it is not used.

k0ekk0ek replied:
Removed it

wcawijngaards replied:
Okay, (I think the restart, restart-.. are also mentioned in the readme for this section, by the way).

k0ekk0ek replied:
Done

Commit:
Remove superfluous assignment

k0ekk0ek commented:
> @k0ekk0ek I'm doing PRs on your branch for you to process (and merge if you like them).
> All PR's have master already merged. Safes you some work. There are currently 2 PR's.
> 
>     1. [k0ekk0ek#2](https://github.com/k0ekk0ek/nsd/pull/2) : Rename VERIFY_IP_PORT into VERIFY_PORT
> 
>     2. [k0ekk0ek#3](https://github.com/k0ekk0ek/nsd/pull/3) : Fix cutest program to use fd's with popen3
>        That last PR should make the cirrus test succeed...

Integrated in my rebase and available once I push. Thanks.

wcawijngaards commented on tpkg/credns-setup/multitail.sh.in:
```diff
@@ -0,0 +1,19 @@
+        DIRNAME=`dirname $file`
+        BASENAME=`basename $DIRNAME`
+	stdbuf -i 0 -o 0 -e 0 tail -f $file \
+      | stdbuf -i 0 -o 0 -e 0 sed "s/nsd\[/$BASENAME[/g" &
```
stdbuf mode `-o L -e L` is better on the second `sed` line, because the logs are line based, and that makes the output line-by-line separated from the different servers, like a syslog output would be.  For the test maybe not that relevant as a bugfix, but for multitail used elsewhere that would be a good fix, I think, unless it is used in a non-line oriented data flow.

k0ekk0ek replied:
Done

wcawijngaards commented on tpkg/credns.tdir/credns.test:
```diff
@@ -0,0 +1,127 @@
+make start-proxy
+wait_nsd_up proxy/nsd.log
+make start-master
+wait_nsd_up proxy/nsd.log
```
Like one before, this needs to be master/nsd.log instead of the second proxy/nsd.log for wait_nsd_up.

k0ekk0ek replied:
Done

k0ekk0ek commented:
Handled the low-hanging fruit and rebased onto `master`. Will now move on to check out the more serious comments

k0ekk0ek commented:
> @k0ekk0ek I am doing some tests with `tpkg/credns-setup` and noticing some issues.
> When testing with just a single zone:
> The verifier taking longer to verify than the SOA Retry timer, results in a constant stream of AXFR queries from the proxy to the master (which are luckily aborted by the querier/proxy):
> 
> ```
> [2020-04-06 11:50:10.903] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
> [2020-04-06 11:50:29.921] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
> [2020-04-06 11:50:47.938] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
> [2020-04-06 11:51:06.957] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
> [2020-04-06 11:51:25.973] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
> [2020-04-06 11:51:44.991] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
> [2020-04-06 11:52:03.009] nsd[31960]: info: axfr for credns.test. from 127.0.0.1
> ```
> 
> This constant stream of AXFR queries eventually stops when the SOA Expire timer runs out.
> 
> Maybe, when the SOA Retry timer goes, we have to check whether the zone is still being verified and then postpone for another SOA Retry seconds?

@wtoorop, do you remember if the `verifier` is killed every time? It may have to do with verification code marking the zone as `is_bad`. That should logically result in an AXFR, I believe that was already the case. Of course, this is undesirable, so we may want to opt for a different state and communicate that back to xfrd.

wtoorop commented:
@k0ekk0ek No the `verifier` is not killed, it is just taking long time resulting in a flood of AXFR's for th proxy to the master. It is pretty easy to reproduce. Do you want me to create a tpkg for it?

wcawijngaards commented on nsd-control.8.in:
```diff
@@ -35,8 +35,11 @@ There are several commands that the server understands.
+default location relative to the installation prefix specified at
+compile-time.  The executable location can be overridden by setting
+\fINSD_PATH in the environment.  It is started with the config file
+specified using \fI\-c\fR or the default config file.
```
You need to en the `\fI` on NSD_PATH with `\fR` after NSD_PATH.  Github renders the italics on the manpage nroff, by the way.

k0ekk0ek replied:
done

wcawijngaards commented on nsd.conf.sample.in:
```diff
@@ -248,6 +248,38 @@ server:
+	# port: 5347
+
+	# Interfaces to bind for zone verification (default are the wildcard
+	# interfaces 127.0.0.1 and ::1). To bind to to multiple IP addresses,
```
s/wildcard//, the localhost addresses are not called wildcard.

k0ekk0ek commented:
@wtoorop, the AXFR flood issue is solved (though I pushed a bit prematurely as some other tests are failing). Previously the `soa_nsd_acquired` field would always be updated if an update was received with a serial that previously been encountered. Since that is not true anymore and the most recent serial is checked (either `soa_xfr_acquired` or `soa_nsd_acquired`, I changed `xfrd_set_timer_refresh` to use the acquired time of the last transfer or active zone, whichever is more recent. The `soa_nsd_acquired` is updated after the reload succeeded.

k0ekk0ek commented:
@wtoorop, I think both issues you mentioned are fixed. At least the AXFR flood. I simplified handling of the xfr packet, consequently I think the issue with the secondary not starting an xfr is also solved. I'm going to verify myself (of course), but if you can do a quick run too, that'd be appreciated.

Commit:
Update xfrd to keep zone transfers until reload succeeds

Keep track of individual zone transfers so that correct updates can be
reapplied if a reload is unsuccessful.

wcawijngaards commented on tpkg/verify.tdir/verify.pre:
```diff
@@ -0,0 +1,32 @@
+NSD_VERIFIER="$(realpath $(dirname ${0}))/verifier.script"
+
+dir=$(realpath $(dirname "$0"))
+
```
`dir` is an unused variable. This line can be removed from the script.

wcawijngaards commented on tpkg/verify_no.tdir/verify_no.pre:
```diff
@@ -0,0 +1,28 @@
+NSD_VERIFIER="$(realpath $(dirname ${0}))/verifier.script"
+
+dir=$(realpath $(dirname "$0"))
+
```
Also unused here, the `dir` line can be removed

wcawijngaards commented on tpkg/verify_no.tdir/verify_no.test:
```diff
@@ -0,0 +1,16 @@
+PRE="../.."
+NSD="${PRE}/nsd"
+
+${NSD} -c $(pwd)/nsd.conf -p ${NSD_PORT} -V 5 -L 5 -F 0xFFFF
```
This test also needs to check for presence of the -L option and if not skip the test.

wcawijngaards commented on tpkg/verify.tdir/verify.test:
```diff
@@ -0,0 +1,60 @@
+PRE="../.."
+NSD="${PRE}/nsd"
+
+${NSD} -c $(pwd)/nsd.conf -p ${NSD_PORT} -V 5 -L 5 -F 0xFFFF
```
This test also needs to check for presence of the -L option and if not skip the test.

wcawijngaards commented on tpkg/verify_timeout.tdir/verify_timeout.pre:
```diff
@@ -0,0 +1,35 @@
+NSD_VERIFIER="$(realpath $(dirname ${0}))/verifier.script"
+NSD_VERIFIER_TIMEOUT=2 # 2 seconds
+
+dir=$(realpath $(dirname "$0"))
```
The variable `dir` is not used.  This line can be removed.

wcawijngaards commented on tpkg/verify_timeout.tdir/verify_timeout.pre:
```diff
@@ -0,0 +1,35 @@
+
+dir=$(realpath $(dirname "$0"))
+
+echo "TESTNS_PORT=${LDNS_PORT}" >> .tpkg.var.test
```
s/LDNS_PORT/TESTNS_PORT/
Fix variable, the LDNS_PORT is not set.

wcawijngaards commented on tpkg/verify.tdir/verify.pre:
```diff
@@ -0,0 +1,32 @@
+
+dir=$(realpath $(dirname "$0"))
+
+echo "TESTNS_PORT=${LDNS_PORT}" >> .tpkg.var.test
```
s/LDNS_PORT/TESTNS_PORT/
Fix variable name, the LDNS_PORT variable is not set.

wcawijngaards commented on tpkg/verify_timeout.tdir/verify_timeout.test:
```diff
@@ -0,0 +1,22 @@
+PRE="../.."
+NSD="$PRE/nsd"
+
+${NSD} -c nsd.conf -L 5 -F 0xFFFF -p ${NSD_PORT}
```
Test if the `-L` option is available.  Like in other tests that need the debug functionality.

wcawijngaards commented on verify.c:
```diff
@@ -0,0 +1,563 @@
+	if (verifier->error_stream.fd != -1) {
+		verify_handle_stream(
+			verifier->error_stream.fd, EV_READ, verifier);
+		verifier->output_stream.fd = -1;
```
s/output_stream/error_stream/
But it does not matter because the routine should have set it to -1, but it is still fine to set it here in case it hasn't.  It would then also be a good idea to close that fd here, if you think it is not always -1.

k0ekk0ek replied:
done

wcawijngaards commented on xfrd.c:
```diff
@@ -2377,77 +2511,106 @@ find_same_master_notify(xfrd_zone_type* zone, int acl_num_nfy)
+			xfr = zone->latest_xfr;
+			while(xfr) {
+				if(!xfr->acquired)
+					continue;
```
This is an infinite loop if !xfr->acquired, it needs to go to the xfr->prev before continue.

k0ekk0ek replied:
(l)oops (sorry, couldn't resist :sweat_smile:, thanks, will fix)

k0ekk0ek replied:
done

wcawijngaards commented on xfrd.c:
```diff
@@ -2377,77 +2511,106 @@ find_same_master_notify(xfrd_zone_type* zone, int acl_num_nfy)
+		while(xfr && xfr->acquired) {
+			/* skip updates that arrived after failed reload */
+			if(xfrd->reload_cmd_first_sent && !xfr->sent)
+				continue;
```
This is also an infinite loop if the condition is true, if !xfr->sent, it should go to xfr->next before continue.

k0ekk0ek replied:
done, should've been a `break`

wcawijngaards reviewed (approved):
Looks good, after I looked through it.

Commit:
Add verify_repat test to ensure verifier options can be reconfigured

Commit:
Make verifier options reconfigurable

Commit:
Fix help return code in nsd-control-setup script

k0ekk0ek commented:
Thanks @wcawijngaards, it's very much appreciated :+1:

wtoorop commented:
Wishlist item: In mixed bad/good zone reloads, do not call verifier again on zones that were already verified.

Commit:
Use retry interval to schedule refresh if zone fails verification

An immediate refresh can potentially trigger an update loop if a zone fails
verification because the primary is (probably) still serving same zone.

Commit:
Add ixfr_gone test to ensure failed update check procedure works

Commit:
Fix handling of SOAINFO if update is received during reload

Commit:
Mark diff files as verified to skip verification the second time

Commit:
Add verify_again test to ensure verified diffs are not verified again

k0ekk0ek commented:
> Wishlist item: In mixed bad/good zone reloads, do not call verifier again on zones that were already verified.

@wtoorop, verifier is not called again. `verify_again` test was added to *verify* :wink:

Commit:
Update ixfr_gone for changes to server reload operation

Commit:
Update reload to communicate soa information only once

Reloads can contain multiple incremental updates. Each must successfully apply
in order for the resulting zone to be correct. Before, if a file was missing,
it would go unnoticed. Corrupt updates could preceed correct updates, resulting
in soa information still being communicated as if changes were successful and
subsequent corrupt or inconsistent updates would invalidate preceding correct
updates too, resulting in both xfrd nsd having a different view.

Commit:
Fix bugs in repattern changes for zone verification

Commit:
Update credns test to ignore duplicate serials

wtoorop commented:
closing in favor of #192
